### PR TITLE
Basic Kafka Producer app

### PR DESCRIPTION
**Pull request by AparnaKarve**
Get the basic producer app running...

- Writes data to S3 bucket
- Sends json data on the bus

`aiops-publisher`  log excerpts of interest -

----------------------------

Connecting to AWS and writing data to S3 -

screen shot 2018-11-01 at 7 43 56 am

Sending json data to the message bus -

screen shot 2018-11-01 at 7 46 16 am

-----------------------------

`aiops-incoming-listener` receiving the json data -

screen shot 2018-11-01 at 7 50 30 am

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -4,6 +4,10 @@ verify_ssl = true
 name = "pypi"
 
 [packages]
+app = "*"
+flask = "*"
+kafka-python = "*"
+"s3fs" = "*"
```
Comment (tumido) [inline Pipfile]: `gunicorn` is missing here
```suggestion
"gunicorn" = "*"

```

File: producer/producer.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+from kafka import KafkaProducer, producer
+
+import json
+
+def publish_message(server: str, topic: str, data: dict) -> dict:
+    """Produces a message with a topic on the message bus"""
```
Comment (tumido) [inline producer/producer.py]: Dot at the end of the first line of the docstring is missing.

File: s3/s3.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+from s3fs import S3FileSystem
+
+
+def connect(key: str, secret: str) -> S3FileSystem:
+    """Create a Boto3 client for S3 service."""
+    filesystem = S3FileSystem(key=key, secret=secret)
+    return filesystem
+
+def save_data(filesystem: str, bucket: str, data: str) -> dict:
+    """saves the data received from AI Service"""
```
Comment (tumido) [inline s3/s3.py]: Capital at the beginning of the docstring and dot at the end :wink:  Nit picks, I know...

File: s3/s3.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    with filesystem.open(bucket, 'w') as f:  #bucket = 'myaiops/aidata'
```
Comment (tumido) [inline s3/s3.py]: Please remove that comment at the EOL.

File: wsgi.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    aws_key = os.environ.get('AWS_ACCESS_KEY_ID')
+    aws_secret = os.environ.get('AWS_SECRET_ACCESS_KEY')
+    aws_bucket = os.environ.get('AWS_S3_BUCKET_NAME')
+
+    application.logger.info('Saving data to bucket %s', aws_bucket)
+    filesystem = s3.connect(aws_key, aws_secret)
+    s3.save_data(filesystem, aws_bucket, "Data going to bucket")
+
+    value = {'url': "INPUT DATA ORIGIN IN S3", 'rh_account': "ACCOUNT"}
+    data_to_send = {'value': value}
```
Comment (tumido) [inline wsgi.py]: You don't need to pass this in the "value" property, pass the `value` dict directly.

File: wsgi.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+    aws_key = os.environ.get('AWS_ACCESS_KEY_ID')
+    aws_secret = os.environ.get('AWS_SECRET_ACCESS_KEY')
+    aws_bucket = os.environ.get('AWS_S3_BUCKET_NAME')
+
+    application.logger.info('Saving data to bucket %s', aws_bucket)
+    filesystem = s3.connect(aws_key, aws_secret)
+    s3.save_data(filesystem, aws_bucket, "Data going to bucket")
+
+    value = {'url': "INPUT DATA ORIGIN IN S3", 'rh_account': "ACCOUNT"}
```
Comment (tumido) [inline wsgi.py]: Let's use our custom format instead. Don't place random strings in random fields. Let's rather do something like:
```suggestion
    value = {
        'message': 'Data published to s3',
        'origin': 'AI-Ops',
        'url': f's3:/{aws_bucket}/<SOME_DESCRIPTION_PLEASE_REPLACE>'
    }
```

File: wsgi.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    data_to_send = {'value': value}
+
+    application.logger.info('Publishing message on topic %s', topic)
+    producer.publish_message(server, "available", data_to_send)
+    return 'aiops-publisher activated!'
```
Comment (tumido) [inline wsgi.py]: Let's return JSON instead (import jsonify from Flask)

```suggestion
    return jsonify(messageg='aiops-publisher activated!')
```

File: wsgi.py (hunk 6 above)
Comment (AparnaKarve) [reply wsgi.py]: I wonder if we need `<SOME_DESCRIPTION_PLEASE_REPLACE>` ...

`f's3:/{aws_bucket}'` should suffice for the url, right?

**Comment by AparnaKarve** [comment]
I have addressed all your feedback, but have a question here - https://github.com/ManageIQ/aiops-publisher/pull/1#discussion_r230176284

**Pull request merged**